Spring security

newoo opened the pull request:
월요일에 진행하지 못해, 오늘 PR 올립니다.

### 금일 진행 사항 
  - jwt인증을 위한 인증 필터 객체 추가 및 구현

### 의존성 추가사항

`implementation 'org.springframework.boot:spring-boot-starter-security'`

Authentication 작업 처리를 위해 추가합니다.
참고 [github](https://github.com/newoo/til/blob/master/spring/spring-boot-starter-security.md)

newoo committed:
Gradle 프로젝트에 spring-boot-starter-security 추가하기

Authentication 작업 처리를 위해 추가

newoo committed:
Spring Security에 대한 설정을 위한 클래스 추가

newoo committed:
jwt를 위한 인증 필터 클래스 추가 및 인증 설정 클래스에 의존성 추가

newoo committed:
jwt를 위한 인증 필터 클래스에 인증요청을 처리하는 객체 의존성 주입

newoo committed:
jwt를 위한 인증 필터 객체생성하는 곳에 인증요청을 처리하는 객체 추가

newoo committed:
인증 인터셉터 내, 인증 구현을 인증 필터 객체로 이동

johngrib commented on app/src/main/java/com/codesoom/assignment/config/SecurityJavaConfig.java:
```diff
@@ -0,0 +1,28 @@
+import javax.servlet.Filter;
+
+/**
+ * Spring Security에 대한 설정
```
이렇게 해도 충분할 것 같아요.
```suggestion
 * Spring Security 설정.
```

johngrib commented on app/src/main/java/com/codesoom/assignment/filters/JwtAuthenticationFilter.java:
```diff
@@ -0,0 +1,75 @@
+import java.io.IOException;
+
+/**
+ * Jwt 인증을 위한 필터 클래스
```
이렇게 하면 어떨까요? 
```suggestion
 * JWT 인증 필터.
```

johngrib reviewed (commented):
이번주도 힘을 내서 학습해 봅시다.

johngrib reviewed (commented):
커밋 메시지를 더 심플하게 작성해도 좋습니다. 짧고. 간결하게.

다음 두 메시지를 비교해 봅시다. 9fdecf0bb49271e56876f8160e78d0b118589ab0
- `Spring Security에 대한 설정을 위한 클래스 추가`
- `Spring Security 설정 클래스 추가`

newoo committed:
인증 에러 필터 생성 및 Spring Security 설정에 추가

newoo committed:
클래스 주석 변경 및 불필요한 import 제거

newoo committed:
경로, http method를 기준으로 필터하는 메소드 JavaDoc 추가

newoo committed:
인터셉터 패키지 삭제

newoo committed:
웹 자바 설정 클래스 삭제

newoo committed:
컨트롤러 에러 어드바이스 클래스 내 불필요한 메서드 삭제

유효하지 않는 토큰 예외 핸들링을 더이상 필요없기 때문에 삭제

newoo committed:
인증토큰을 활용하여 인증필터 내부구현

인증토큰 추가에 따라 제품 컨트롤러의 생성 메서드에 인자값 추가

newoo committed:
유저 인증 토큰 클래스 생성

유저 인증 여부를 확인하기 위해 생성

newoo committed:
경로, http method를 기준으로 필터하는 객체 삭제

추가적으로 다른 객체에서의 해당 객체 의존성 제거

newoo committed:
PreAuthorize를 활용하여 메소드 호출 허용 결정

newoo committed:
유저 인증 토큰 isAuthenticated 메서드 오버라이드

테스트실행 시, Null authentication문제로 인한 테스트 실패를 해결하기 위해 추가

newoo committed:
PreAuthorize를 상품 컨트롤러의 수정, 삭제 메소드에도 추가

newoo committed:
상품 컨트롤러 삭제 테스트 추가

토큰이 없을 경우에 대한 삭제 테스트 추가

newoo committed:
유저 권한을 부여하는 기능 추가

newoo committed:
참고사항 주석 추가

newoo committed:
세션 매니지먼트 추가

인증된 사용자와 관련된 세션을 추적하기 위해 추가

newoo commented:
커밋 내용을 짧게 적으려고 고민해보기는 했지만, 생각만큼 잘 되지 않네요.
남은 기간동안에 좀 더 깊은 고민을 해보겠습니다.

오늘은 남은 시간에 금일 추가한 기능에 대해 좀 더 깊이 알아보는 시간을 가지려고 합니다.

### 금일 진행 사항
  - 인증 필터 기능 완성
  - 부족한 테스트 추가

newoo commented on app/src/main/java/com/codesoom/assignment/config/SecurityJavaConfig.java:
```diff
@@ -0,0 +1,41 @@
+                .sessionCreationPolicy(SessionCreationPolicy.STATELESS)
+                .and()
+                .exceptionHandling()
+                .authenticationEntryPoint(new HttpStatusEntryPoint(HttpStatus.UNAUTHORIZED));
```
체인이 길게 되어있다면, 중간 중간 주석을 추가하는 것이 좋을까요?
(위처럼 함수형 스타일로 길게 늘여진 상태를 체인이라고 부르는 것이 정확한지는 모르겠습니다.)

johngrib replied:
필요하다면 주석을 넣어주는 것도 괜찮아요.

johngrib reviewed (commented):
b2b737857b86f4dd46cc813e5f46d6ecb3b3eeb5 커밋 메시지 본문에 작업의 의미에 대한 설명이 있었다면 더 좋았을 거에요.

johngrib reviewed (commented):
8314915b62655164a015c7ad0f193543514a136a `클래스 주석 변경 및 불필요한 import 제거` 이런 커밋 메시지는 다음과 같이 작성했다면 어땠을까요?

`클래스 주석을 읽기 좋게 고치고, 불필요한 import 제거`

johngrib reviewed (commented):
eabdd2c25b3f6c0e69a3cc88043f0018c29aec3b 이런 커밋은 다른 JavaDoc 작업과 함께 커밋해도 좋겠죠.

johngrib reviewed (commented):
43744881cdfbbe73f79f01741cbc7e15391f41d1 

![image](https://user-images.githubusercontent.com/1855714/110788021-6a596d80-82b1-11eb-8ae0-c5527d5f8973.png)

`컨트롤러 에러 어드바이스 클래스 내 불필요한 메서드 삭제`를 다음과 같이 작성했다면 어떨까요?

`유효하지 않은 토큰 예외 발생시 UNAUTHORIZED로 응답하는 예외 핸들링을 삭제하라`

johngrib commented on app/src/main/java/com/codesoom/assignment/config/SecurityJavaConfig.java:
```diff
@@ -5,15 +5,19 @@
  * Spring Security 설정.
  */
 @Configuration
+@EnableGlobalMethodSecurity(prePostEnabled = true)
```
이 애노테이션이 어떤 의미가 있는지 커밋 메시지 본문에 작성해주셨으면 좋았을 거에요.

johngrib commented on app/src/main/java/com/codesoom/assignment/filters/AuthenticationErrorFilter.java:
```diff
@@ -0,0 +1,24 @@
+import javax.servlet.http.HttpServletResponse;
+import java.io.IOException;
+
+public class AuthenticationErrorFilter extends HttpFilter {
```
JavaDoc 주석을 작성해 주세요.

johngrib commented on app/src/main/java/com/codesoom/assignment/security/UserAuthentication.java:
```diff
@@ -0,0 +1,50 @@
+import java.util.List;
+
+/**
+ * 유저 인증 토큰.
+ * 유저 인증 여부를 확인하기 위해 사용되는 객체.
```
주석을 더 심플하게 작성해도 괜찮을 것 같아요.

newoo committed:
인증 에러 필터 짧은 JavaDoc추가

newoo committed:
토큰없이 유저정보 수정 요청 테스트 추가

newoo committed:
토큰없이 유저정보 수정 요청 테스트 통과

그 외 유저정보 수정 요청 테스트 실패

newoo committed:
유저정보 수정 요청 전체 테스트 통과

인증과정 통과를 위해  전체 테스트 MockMVC에 토큰값 추가

newoo committed:
유효하지 않는 토큰으로 유저정보 수정 요청하는 테스트 추가

newoo committed:
유효하지 않는 토큰으로 유저정보 수정 요청하는 테스트 통과

유저 컨트롤러에 인증 서비스 의존성 추가
유저 컨트롤러 테스트에 인증 서비스 관련 stub 추가

newoo committed:
토큰없이 유저정보 삭제 요청 테스트 추가

newoo committed:
토큰없이 유저정보 삭제 요청 테스트 통과

newoo committed:
유효하지 않는 토큰으로 유저정보 삭제 요청하는 테스트 추가 후 통과

newoo commented:
UserController관련 테스트 중 수정, 삭제 테스트에 인증 관련 조건을 추가했습니다.
일단 먼저 PR 업데이트 후 계속 진행하겠습니다.
JavaDoc과 커밋도 좀 더 고민해보겠습니다.

johngrib reviewed (commented):
cb3637654fc2cde300b4daab03d662d4a2f0ba7d `짧은`은 불필요합니다.
![image](https://user-images.githubusercontent.com/1855714/110939400-b4f1ed00-8378-11eb-8e13-f6fcaa17c3a6.png)

johngrib reviewed (commented):
이 두 커밋은 하나의 커밋이었으면 더 적절했을 거에요. 커밋이 너무 자잘하게 나뉘어 있어서 관리가 불편해집니다.
21245b7 c04b08a

johngrib reviewed (commented):
`git rebase -i`의 `fixup`과 `squash`에 대해 조사해 보시고, 연습도 해 보세요.

newoo committed:
유저 필드에 기본값 지정

newoo committed:
유저의 비밀번호 암호화 테스트 추가하고, 암호화 메서드 추가

newoo committed:
유저등록 시, 비밀번호를 암호화되도록 변경

newoo committed:
유저 인증 테스트 2곳에 비밀번호 암호화 과정 추가

유저 등록 시, 비밀번호 암호화 기능이 추가됨에 따라
평문의 비밀번호를 비교해서는 테스트 통과가 불가능하기 때문에
비밀번호 암호화 과정을 추가

newoo committed:
유저 인증 메소드 내부 구현 변경

기존의 평문 비밀번호가 일치 여부를 판단했던 방법에서
비밀번호 암호화에 따라 암호화한 비밀번호와 매치되는지 여부로 변경

newoo committed:
유저 인증, 비밀번호 메소드에 암호화 인코더 의존성 주입

암호화 인코더 생성 책임을 유저 객체가 아닌 유저를 생성하는 객체에게 위임
부수적으로 테스트통과를 위해 인증 서비스와 유저 서비스에도 의존성 주입

newoo committed:
인증 서비스 로그인 메소드 변경

내부 구현에 임의의 유저 아이디으로 지정된 것을 실제 유저 아이디로 변경

newoo committed:
유저가  다른유저의 정보를 수정 요청하는 테스트 추가

newoo committed:
비허가 접근 예외 추가

newoo committed:
비허가 접근에 대한 에러 응답 추가

newoo commented:
아직 git rebase -i의 fixup과 squash에 대해 조사해보지는 못했습니다.
오늘 밤과 내일 마무리 작업을 진행하면서 조사해보겠습니다.

자체 테스트는 전체 통과했으나, API 테스트 1개를 아직 통화 못 한 상황입니다.
`다른 유저가 유저정보를 변경하려고 할 때, 응답의 http status가 403이어야하는 테스트`입니다.

컨트롤러나 서비스에서 핸들링을 해주어야할 것으로 생각되어서 작업을 진행했다가
현재 진행사항들을 reset한 상황입니다.

스프링 자체적으로 지원하는 기능이 있는지도 검토해보려고 합니다.

johngrib commented on app/src/main/java/com/codesoom/assignment/domain/User.java:
```diff
@@ -19,18 +19,20 @@
     private Long id;
 
-    private String email;
+    @Builder.Default
+    private String email = "";
 
-    private String name;
+    @Builder.Default
+    private String name = "";
 
-    private String password;
+    @Builder.Default
+    private String password = "";
```
이렇게 기본값을 지정한 이유에 대해 차분하게 설명해 보세요. 커밋 메시지를 사용하셔도 좋고요, PR에 리뷰 코멘트를 작성하셔도 좋습니다.

johngrib commented on app/src/main/java/com/codesoom/assignment/domain/User.java:
```diff
@@ -42,4 +44,9 @@ public void destroy() {
         return !deleted && password.equals(this.password);
     }
+
+    public void changePassword(String password) {
```
여기에 JavaDoc 주석을 작성해주세요.

johngrib reviewed (commented):
55f002a2f4d3850a70ef59838f0f20d893754958 커밋 메시지가 아쉽습니다.

johngrib reviewed (commented):
80813d6 커밋 메시지 본문의 문장이 이해하기 어렵습니다.

![image](https://user-images.githubusercontent.com/1855714/111029672-462c9680-8441-11eb-9c37-89f71dfc2122.png)

다음과 같이 작성하면 어떨까요?

```
패스워드 암호화 기능이 추가되면서 실패하게 된 테스트를 교정했습니다.
이번에 고친 테스트는 패스워드를 평문으로 사용하고 있었습니다.
```

johngrib reviewed (commented):
98466cf17a4d344003c21e506ec46be1d322358d 커밋 메시지가 어렵습니다.

![image](https://user-images.githubusercontent.com/1855714/111029776-d10d9100-8441-11eb-9133-38bd1be28b42.png)

다음과 같이 수정하면 어떨까요?

```
패스워드 검증에 암호화 적용

사용자 인증 과정에서 맞는 패스워드인지 확인할 때 평문을 사용하고 있었습니다.
그래서 암호화된 패스워드를 사용해 맞는 패스워드인지 확인하도록 고쳤습니다.
```

johngrib commented on app/src/main/java/com/codesoom/assignment/errors/UnauthorizedAccessException.java:
```diff
@@ -0,0 +1,11 @@
+package com.codesoom.assignment.errors;
+
+/**
+ * 비허가 접근 예외
```
JavaDoc description에는 마침표를 꼭 찍어주세요.

newoo committed:
비허가 접근에 대한 테스트 통과

인증 객체를 이용해 수정하려는 유저가 동일 유저인지 판단

newoo committed:
불필요한 import 제거

newoo committed:
코드 커버리지 문제로 롬복 설정 추가

newoo commented:
직접 테스트할 때는 정상적으로 기능이 동작하는데, 로컬에서도 E2E 테스트를 진행봤지만, 특정 2개의 테스트가 계속 실패합니다;;
금일 저녁에 개인사정때문에 더 이상 과제 진행이 어려워 금주 과제는 일단 여기서 마무리 하겠습니다...

개인 일정 마무리 후, git rebase -i의 fixup과 squash에 대해 정리해보고 TIL이나 개인 블로그에 정리해보려고 합니다.

1주일동안 피드백 감사합니다. 👍 

이번주 회고입니다.
[링크](https://oen0609.tistory.com/56)

johngrib reviewed (commented):
한 주간 고생 많으셨습니다.
